src/Interpreters/ExpressionActions.cpp

```diff
@@ -68,21 +70,100 @@ ExpressionActionsPtr ExpressionActions::clone() const
     return std::make_shared<ExpressionActions>(*this);
 }
 
+void ExpressionActions::rewriteShortCircuitArguments(const ActionsDAG::NodeRawConstPtrs & children, const std::unordered_map<const ActionsDAG::Node *, bool> & need_outside, bool force_rewrite)
+{
+    for (const auto * child : children)
+    {
+        /// Skip actions that are needed outside or have already been rewritten.
+        if (!need_outside.contains(child) || need_outside.at(child) || child->lazy_execution != ActionsDAG::LazyExecution::DISABLED)
+            continue;
+
+        switch (child->type)
+        {
+            case ActionsDAG::ActionType::FUNCTION:
+                rewriteShortCircuitArguments(child->children, need_outside, force_rewrite);
+                const_cast<ActionsDAG::Node *>(child)->lazy_execution = force_rewrite ? ActionsDAG::LazyExecution::FORCE_ENABLED : ActionsDAG::LazyExecution::ENABLED;
+                break;
+            /// Propagate lazy execution through aliases.
+            case ActionsDAG::ActionType::ALIAS:
+                rewriteShortCircuitArguments(child->children, need_outside, force_rewrite);
+                break;
+            default:
```

Comment:
I suppose `arrayJoin` also should be supported.
E.g.: `select if(number != 0, arrayJoin([1, intDiv(42, number), 2]), 0)`
Here, arrayJoin for short-circuit column will work as if array is empty for `mask = 0`.

Reply:
> I suppose arrayJoin also should be supported

I spent a lot of time thinking about how to support it for arrayJoin, but I considered that it's almost impossible. I wrote a comment about it:
https://github.com/ClickHouse/ClickHouse/blob/591cecc66499e99948001a82b0f03e1ced4c7763/src/Interpreters/ExpressionActions.cpp#L81-L87

Reply:
But I think it is ok to get offsets after filtering.
`select if (number % 2 = 0, [], arrayJoin(range(number))) from numbers(4);`
Ordinary evaluation
`[0, 1, 2, 3] -> range -> [[], [0], [0, 1], [0, 1, 2]] -> arrayJoin -> [0, 0, 1, 0, 1, 2] -> if -> [0, 0, 1, 2]`
`[1, 0, 1, 0] ->                                                       arrayJoin -> [0, 1, 1, 0, 0, 0]`

short-circuit evaluation
`if -> [?, 1, ?, 3] -> range -> [?, [0], ?, [0, 1, 2]] -> arrayJoin -> [0, 0, 1, 2]`
`if -> [1, 0, 1, 0]`

But we can support it later (if needed)